### PR TITLE
create spotMetaAndAssetCtxs function

### PR DESCRIPTION
**Pull request by RaduGeorgescu**
created and added a function for spot_meta_and_asset_ctxs
as per the documentation
https://hyperliquid.gitbook.io/hyperliquid-docs/for-developers/api/info-endpoint/spot#retrieve-spot-asset-contexts

### CODE REVIEW

File: hyperliquid/utils/types.py [hunk 1]
```diff
@@ -29,6 +29,14 @@
 )
 SpotMeta = TypedDict("SpotMeta", {"universe": List[SpotAssetInfo], "tokens": List[SpotTokenInfo]})
 
+
+SpotdAssetCtxsTokenInfo = TypedDict("SpotdAssetCtxsTokenInfo", {"name": str, "szDecimals": int, "weiDecimals": int})
```
Comment (traderben) [inline hyperliquid/utils/types.py:33]: This can just be SpotTokenInfo from above right?

File: hyperliquid/utils/types.py [hunk 2]
```diff
@@ -29,6 +29,14 @@
...
+SpotdAssetCtxsUniverseInfo = TypedDict("SpotdAssetCtxsUniverseInfo", {"name": str, "tokens": List[int]})
+SpotdAssetCtxsMarketDataInfo = TypedDict("SpotdAssetCtxsMarketDataInfo", {"dayNtlVlm": str, "markPx": str, "midPx": str, "prevDayPx": str})
+
+SpotMetaAndAssetCtxsDict = TypedDict("SpotMetaAndAssetCtxsDict", {"tokens": List[SpotdAssetCtxsTokenInfo], "universe": List[SpotdAssetCtxsUniverseInfo]})
+SpotMetaAndAssetCtxs = List[Union[SpotMetaAndAssetCtxsDict, List[SpotdAssetCtxsMarketDataInfo]]]
```
Comment (traderben) [inline hyperliquid/utils/types.py:38]: This type can be better, it should just be a tuple of SpotMeta and List[SpotAssetCtx]:
SpotMetaAndAssetCtxs = Tuple[SpotMeta, List[SpotAssetCtx]]

File: hyperliquid/utils/types.py [hunk 3]
```diff
@@ -29,6 +29,14 @@
 )
 SpotMeta = TypedDict("SpotMeta", {"universe": List[SpotAssetInfo], "tokens": List[SpotTokenInfo]})
 
+
+SpotdAssetCtxsTokenInfo = TypedDict("SpotdAssetCtxsTokenInfo", {"name": str, "szDecimals": int, "weiDecimals": int})
+SpotdAssetCtxsUniverseInfo = TypedDict("SpotdAssetCtxsUniverseInfo", {"name": str, "tokens": List[int]})
+SpotdAssetCtxsMarketDataInfo = TypedDict("SpotdAssetCtxsMarketDataInfo", {"dayNtlVlm": str, "markPx": str, "midPx": str, "prevDayPx": str})
+
+SpotMetaAndAssetCtxsDict = TypedDict("SpotMetaAndAssetCtxsDict", {"tokens": List[SpotdAssetCtxsTokenInfo], "universe": List[SpotdAssetCtxsUniverseInfo]})
```
Comment (traderben) [inline hyperliquid/utils/types.py:37]: no need to define this, this is just SpotMeta from above

File: hyperliquid/utils/types.py [hunk 4]
```diff
@@ -29,6 +29,14 @@
 )
 SpotMeta = TypedDict("SpotMeta", {"universe": List[SpotAssetInfo], "tokens": List[SpotTokenInfo]})
 
+
+SpotdAssetCtxsTokenInfo = TypedDict("SpotdAssetCtxsTokenInfo", {"name": str, "szDecimals": int, "weiDecimals": int})
+SpotdAssetCtxsUniverseInfo = TypedDict("SpotdAssetCtxsUniverseInfo", {"name": str, "tokens": List[int]})
+SpotdAssetCtxsMarketDataInfo = TypedDict("SpotdAssetCtxsMarketDataInfo", {"dayNtlVlm": str, "markPx": str, "midPx": str, "prevDayPx": str})
```
Comment (traderben) [inline hyperliquid/utils/types.py:35]: name this SpotAssetCtx. Why did you add these lowercase d's?
```suggestion
SpotAssetCtx = TypedDict("SpotAssetCtx", {"dayNtlVlm": str, "markPx": str, "midPx": Optional[str], "prevDayPx": str, "circulatingSupply": str, "coin": str})
```

**Review by traderben — CHANGES_REQUESTED** [review]

File: hyperliquid/utils/types.py (hunk 4 above)
Comment (RaduGeorgescu) [reply hyperliquid/utils/types.py:35]: idk where you got the "circulatingSupply": str, "coin": str part

File: hyperliquid/utils/types.py (hunk 2 above)
Comment (RaduGeorgescu) [reply hyperliquid/utils/types.py:38]: well, I tried to replicate the response which is 
array[json{},array[]]
```
[
{
    "tokens": [
        {
...
        },
        {
...
        }
    ],
    "universe": [
        {
...
        }
    ]
},
[
    {
...
    }
]
]
```

File: hyperliquid/utils/types.py (hunk 3 above)
Comment (traderben) [reply hyperliquid/utils/types.py:37]: Did you try sending the requests and looking at the responses? We probably missed updating some documentation

File: hyperliquid/utils/types.py (hunk 4 above)
Comment (traderben) [reply hyperliquid/utils/types.py:35]: You can look at the returned data

File: hyperliquid/utils/types.py (hunk 2 above)
Comment (traderben) [reply hyperliquid/utils/types.py:38]: The top level result should always have 2 elements the first of which is the SpotMeta and the second one is a list of SpotAssetCtx, if you define it as a list of a union that doesn't specify anything about the length or order of the response

File: hyperliquid/utils/types.py (hunk 3 above)
Comment (RaduGeorgescu) [reply hyperliquid/utils/types.py:37]: I tried just now and you are write. Updated info.py to reflect what comes as a response.

File: hyperliquid/utils/types.py (hunk 2 above)
Comment (RaduGeorgescu) [reply hyperliquid/utils/types.py]: u r right, I implemented the tuple.

File: hyperliquid/utils/types.py (hunk 4 above)
Comment (RaduGeorgescu) [reply hyperliquid/utils/types.py]: I implemented some of the changes

**Review by traderben — APPROVED** [review]
